### PR TITLE
doc/dev: add crimson.rst

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by tchaikov** [comment]
we could fill this document with more topics, like

- how crimson will interact with bluestore
- how we plan to integrate with SPDK
- how the error handling should look like
- the rationales behind the controversial design decisions we made

but i think this is a start.

File: doc/dev/crimson.rst [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
++---------+---------+
+| < 0     | error   |
++---------+---------+
+|   0     | warn    |
++---------+---------+
+| [1, 5)  | info    |
++---------+---------+
+| [5, 20) | debug   |
++---------+---------+
+| >= 20   | info    |
```
Comment (athanatos) [inline doc/dev/crimson.rst]: ?

**Comment by athanatos** [comment]
LGTM other than the typo.

File: doc/dev/crimson.rst (hunk 1 above)
Comment (tchaikov) [reply doc/dev/crimson.rst]: fixed and repushed.

File: doc/dev/crimson.rst [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+crimson is the code name of crimson-osd, which the next generation ceph-osd. It
+targets fast storage devices, like NVMe storage, for better performance.
+
+
+daemonize
+---------
+
+Unlike ``ceph-osd``, ``crimson-osd`` does daemonize itself even if the
+``daemonize`` option is enabled. Because, to read this option, ``crimson-osd``
```
Comment (xiexingguo) [inline doc/dev/crimson.rst]: > ``crimson-osd`` does daemonize itself even if the``daemonize`` option is enabled.

FYI, the above sentence confuses me. s/enabled/not enabled/?

File: doc/dev/crimson.rst [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+needs to ready its config sharded service, but this sharded service lives
```
Comment (xiexingguo) [inline doc/dev/crimson.rst]: ready or read?

Comment (tchaikov) [reply doc/dev/crimson.rst]: see https://www.merriam-webster.com/dictionary/ready#h2-anchor

actually both should work, but if "read" is used, probably i would go with "read from".

File: doc/dev/crimson.rst (hunk 2 above)
Comment (tchaikov) [reply doc/dev/crimson.rst]: yeah, it is confusing. i put `-f` but changed my mind and used `daemonize` instead. but failed to update the first part of the sentence.

**Comment by tchaikov** [comment]
@xiexingguo updated and repushed.

File: doc/dev/crimson.rst (hunk 3 above)
Comment (xiexingguo) [reply doc/dev/crimson.rst]: Ah, sorry. Indeed.

File: doc/dev/crimson.rst [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+daemonize
+---------
+
+Unlike ``ceph-osd``, ``crimson-osd`` does not daemonize itself even if the
+``daemonize`` option is enabled. Because, to read this option, ``crimson-osd``
+needs to ready its config sharded service, but this sharded service lives
+in the seastar reactor. If we fork a child process and exit the parent after
+starting the Seastar engine, that will leave us with a single thread which is
+the replica of the thread calls `fork()`_. This would unnecessily complicate
```
Comment (cyx1231st) [inline doc/dev/crimson.rst]: s/unnecessily/unnecessarily

File: doc/dev/crimson.rst [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+does not send the logging message to specified ``log_file``. It writes
+the logging messages to stdout and/or syslog. Again, this behavior can be
+changed using ``--log-to-stdout`` and ``--log-to-syslog`` command line
+options. By default, ``log-to-stdout`` is enabled, and the latter disabled.
+
+
+vstart.sh
+---------
+
+To facilidate the development of crimson, following options would be handy when
```
Comment (cyx1231st) [inline doc/dev/crimson.rst]: s/facilidate/facilitate

File: doc/dev/crimson.rst [hunk 6]
```diff
@@ -0,0 +1,143 @@
+=======
+crimson
+=======
+
+crimson is the code name of crimson-osd, which the next generation ceph-osd. It
+targets fast storage devices, like NVMe storage, for better performance.
```
Comment (cyx1231st) [inline doc/dev/crimson.rst]: Currently the scope is (block-addressable) NVMe SSD and (byte-addressable) persistent memory, right?

Comment (tchaikov) [reply doc/dev/crimson.rst]: i think we will also support HDD, and none NVMe SSD.

**Comment by tchaikov** [comment]
@cyx1231st updated and repushed.

**Pull request merged**